src/weights.jl

```diff
@@ -291,3 +291,96 @@ end
 
 wmedian(v::RealVector, w::RealVector) = median(v, weights(w))
 wmedian{W<:Real}(v::RealVector, w::WeightVec{W}) = median(v, w)
+
+
+
+###### Weighted quantile #####
+# Variables m, g, low, high, correspond respectively to m, gamma, x[j], x[j+1] in the R help file for quantile
+
+function Base.quantile{W<:Real}(v::RealVector, p::RealVector, w::WeightVec{W})
+
+    isempty(v) && error("quantile of an empty array is undefined")
+    isempty(p) && throw(ArgumentError("empty quantile array"))
+
+    ppermute = sortperm(p)
+    p = p[ppermute]
+
+    # make sure the quantiles are in [0,1]
+    p = bound_quantiles(p)
+
+    if w.sum == 0
+        error("weight vector cannot sum to zero")
+    end
+    if length(v) != length(w)
+        error("data and weight vectors must be the same size. got $(length(v)) and $(length(w))")
+    end
+    for x in w.values
+        isnan(x) && error("weight vector cannot contain NaN entries")
+        x < 0 && error("weight vector cannot contain negative entries")
+    end
+    for x in v
+        isnan(x) && return x
+    end
+    lv = length(v)
+    lp = length(p)
+    m = 1 - p
+    index = (lv * p + m) * w.sum / lv
+    wt = w.values
+    vpermute = sortperm(v)
+    low = fill(v[vpermute[end]], lp)
+    high = fill(v[vpermute[end]], lp)
+    g = fill(0.0, lp)
+    cumulative_weight = zero(eltype(wt))
```

Comment:
Actually, it just occurred to me that a perfect type for this would be `typeof(w.sum)`, as by definition it's known to be able to hold the sum of all weights.